test/Npgsql.Tests/Replication/PgOutputReplicationTests.cs

```diff
@@ -269,24 +310,73 @@ public  Task UpdateForFullReplicaIdentity()
                     var transactionXid = await AssertTransactionStart(messages);
 
                     // Relation
-                    var relMsg = await NextMessage<RelationMessage>(messages);
-                    Assert.That(relMsg.TransactionXid, IsStreaming ? Is.EqualTo(transactionXid) : Is.Null);
-                    Assert.That(relMsg.RelationReplicaIdentitySetting, Is.EqualTo('f'));
-                    Assert.That(relMsg.Namespace, Is.EqualTo("public"));
-                    Assert.That(relMsg.RelationName, Is.EqualTo(tableName));
-                    Assert.That(relMsg.Columns.Count, Is.EqualTo(2));
-                    Assert.That(relMsg.Columns[0].ColumnName, Is.EqualTo("id"));
-                    Assert.That(relMsg.Columns[1].ColumnName, Is.EqualTo("name"));
+                    var relationMsg = await NextMessage<RelationMessage>(messages);
+                    Assert.That(relationMsg.TransactionXid, IsStreaming ? Is.EqualTo(transactionXid) : Is.Null);
+                    Assert.That(relationMsg.ReplicaIdentity, Is.EqualTo(ReplicaIdentitySetting.AllColumns));
+                    Assert.That(relationMsg.Namespace, Is.EqualTo("public"));
+                    Assert.That(relationMsg.RelationName, Is.EqualTo(tableName));
+                    Assert.That(relationMsg.Columns.Count, Is.EqualTo(2));
+                    Assert.That(relationMsg.Columns[0].ColumnName, Is.EqualTo("id"));
+                    Assert.That(relationMsg.Columns[1].ColumnName, Is.EqualTo("name"));
 
                     // Update
                     var updateMsg = await NextMessage<FullUpdateMessage>(messages);
                     Assert.That(updateMsg.TransactionXid, IsStreaming ? Is.EqualTo(transactionXid) : Is.Null);
-                    Assert.That(updateMsg.OldRow!.Length, Is.EqualTo(2));
-                    Assert.That(updateMsg.OldRow!.Span[0].Value, Is.EqualTo("1"));
-                    Assert.That(updateMsg.OldRow!.Span[1].Value, Is.EqualTo("val1"));
-                    Assert.That(updateMsg.NewRow.Length, Is.EqualTo(2));
-                    Assert.That(updateMsg.NewRow.Span[0].Value, Is.EqualTo("1"));
-                    Assert.That(updateMsg.NewRow.Span[1].Value, Is.EqualTo("val1_updated"));
+                    Assert.That(updateMsg.Relation, Is.SameAs(relationMsg));
+
+                    // Current API
+
+                    await foreach (var column in updateMsg.OldRow)
+                    {
+                        // Do old row stuff
+                    }
+
+                    await foreach (var column in updateMsg.NewRow)
+                    {
+                        // Do new row stuff
+                    }
+
+                    // Enumerable API
```

Comment:
@Brar, I pushed the change to publicly expose ReplicationTuple along with its column count.

Re the enumerable multi-row API we discussed in https://github.com/npgsql/npgsql/pull/4064#discussion_r737666799, here's a mock-up of what this would look like (assuming I'm thinking of the same thing as you).

If you're really convinced this is the better option (compared to the current one above), I can do it - but it's hard for me to convince myself it's better :) At the end of the day, it's true that first-time newcomers to the API may get the order wrong above and will get an exception - although it's very likely they'd just copy code out from our docs, where we'd have a correct sample. From that point they shouldn't really have a problem, and end up with clear, concise code. In contrast, the code below stays forever, is hard to understand once you come back to it, and seems "over the top" for the task at hand...

The one exception is if we believe ordering really may change on the PostgreSQL side, in which case the code below makes no assumption. I can't really imagine this happening though.

Reply:
I thought about it one more time and I think the API above is fine.
It deviates from the "pure streaming" approach and has internal assumptions but I guess most people would intuitively get it right and the others would probably read the docs after getting an exception.

Reply:
Thanks @Brar!